packages/excubiae/contracts/extensions/EASExcubia.sol

```diff
@@ -0,0 +1,86 @@
+
+    /// @notice Mapping to track which attestations have been registered by the contract to
+    /// avoid double checks with the same attestation.
+    mapping(bytes32 => bool) public registeredAttestations;
+
+    /// @notice Error thrown when the attestation has been already used to pass the gate.
+    error AlreadyRegistered();
+
+    /// @notice Error thrown when the attestation does not match the designed schema.
+    error UnexpectedSchema();
+
+    /// @notice Error thrown when the attestation does not match the designed trusted attester.
+    error UnexpectedAttester();
+
+    /// @notice Error thrown when the attestation does not match the passerby as recipient.
+    error UnexpectedRecipient();
+
+    /// @notice Error thrown when the attestation has been revoked.
+    error RevokedAttestation();
+
+    /// @notice Constructor to initialize with target EAS contract with specific attester and schema.
+    /// @param _eas The address of the EAS contract.
+    /// @param _attester The address of the trusted attester.
+    /// @param _schema The schema ID that attestations must match.
+    constructor(address _eas, address _attester, bytes32 _schema) {
+        if (_eas == address(0) || _attester == address(0)) revert ZeroAddress();
+
+        EAS = IEAS(_eas);
+        ATTESTER = _attester;
+        SCHEMA = _schema;
+    }
+
+    /// @notice Overrides the `_pass` function to register a correct attestation.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Encoded attestation ID.
+    function _pass(address passerby, bytes calldata data) internal override {
+        super._pass(passerby, data);
+
+        registeredAttestations[decodeAttestationId(data)] = true;
+    }
+
+    /// @notice Overrides the `_check` function to validate the attestation against specific criteria.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Encoded attestation ID.
+    /// @return True if the attestation meets all criteria, revert otherwise.
+    function _check(address passerby, bytes calldata data) internal view override returns (bool) {
+        bytes32 attestationId = decodeAttestationId(data);
+
+        if (registeredAttestations[attestationId]) revert AlreadyRegistered();
+
+        Attestation memory attestation = EAS.getAttestation(attestationId);
+
+        if (attestation.schema != SCHEMA) revert UnexpectedSchema();
+        if (attestation.attester != ATTESTER) revert UnexpectedAttester();
+        if (attestation.recipient != passerby) revert UnexpectedRecipient();
+        if (attestation.revocationTime != 0) revert RevokedAttestation();
+
+        return true;
+    }
+
+    /// @notice Decodes an EAS attestation identifier from the encoded form.
+    /// @param data Encoded attestation ID.
+    /// @return Decoded attestation ID.
+    function decodeAttestationId(bytes calldata data) private pure returns (bytes32) {
```

Comment:
is this needed? what are the benefits vs directly calling `abi.decode` on line #66?

Reply:
if we follow the first suggestion, it's not needed anymore.